pkg/cache/simple.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2017 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/envoyproxy/go-control-plane/api"
+	"github.com/golang/protobuf/proto"
+)
+
+// SimpleCache is a snapshot-based cache that maintains a single version per
+// xDS response, node group tuple, with no canary updates.
+type SimpleCache struct {
+	// responses are cached resources
+	responses map[Key]map[ResponseType][]proto.Message
+
+	// versions must have the same key set as responses
+	versions map[Key]map[ResponseType]int
+
+	// watches keeps track of open watches
+	watches map[Key]map[ResponseType]map[int64]Watch
+
+	// callback requests missing responses
+	callback func(Key, ResponseType)
+
+	// watchCount is the ID generator for watches
+	watchCount int64
+
+	// groups is the hashing function for proxy nodes
+	groups NodeGroup
+
+	mu sync.Mutex
```

Comment:
can we shift to a RW lock? a bit more scalable, given that we have 90% reads by a lot of clients.